### ISSUE TITLE
How do you apply customized loss function?

### ISSUE THREAD

**Issue by jarork** [huggingface/peft#773]
Recently I am working on a tool utilization (mainly calculator) project. (Q: What is 32 times 13? A: The answer is <API>Calculator(32*13)</API>. ) 

In this task, I need the original loss to be timed by a coefficient number for punishing the model outputs if the API calculation results are wrong. What do I need to do with PEFT to apply customized loss function?

Thanks!

**Comment by BenjaminBossan** [comment]
It really depends on how you train your model. PEFT is agnostic with regards to that, so it doesn't impose any kind of loss. It is up to you to define it.

**Comment by jarork** [comment]
> It really depends on how you train your model. PEFT is agnostic with regards to that, so it doesn't impose any kind of loss. It is up to you to define it.

I am using a transformer trainer. I notice the loss is defined here (transformers.trainer_pt_utils)
@dataclass
class LabelSmoother:
    """
    Adds label-smoothing on a pre-computed output from a Transformers model.

    Args:
        epsilon (`float`, *optional*, defaults to 0.1):
            The label smoothing factor.
        ignore_index (`int`, *optional*, defaults to -100):
            The index in the labels to ignore when computing the loss.
    """

    epsilon: float = 0.1
    ignore_index: int = -100

    def __call__(self, model_output, labels, shift_labels=False):
        logits = model_output["logits"] if isinstance(model_output, dict) else model_output[0]
        if shift_labels:
            logits = logits[..., :-1, :].contiguous()
            labels = labels[..., 1:].contiguous()

        log_probs = -nn.functional.log_softmax(logits, dim=-1)
        if labels.dim() == log_probs.dim() - 1:
            labels = labels.unsqueeze(-1)

        padding_mask = labels.eq(self.ignore_index)
        # In case the ignore_index is -100, the gather will fail, so we replace labels by 0. The padding_mask
        # will ignore them in any case.
        labels = torch.clamp(labels, min=0)
        nll_loss = log_probs.gather(dim=-1, index=labels)
        # works for fp16 input tensor too, by internally upcasting it to fp32
        smoothed_loss = log_probs.sum(dim=-1, keepdim=True, dtype=torch.float32)

        nll_loss.masked_fill_(padding_mask, 0.0)
        smoothed_loss.masked_fill_(padding_mask, 0.0)

        # Take the mean over the label dimensions, then divide by the number of active elements (i.e. not-padded):
        num_active_elements = padding_mask.numel() - padding_mask.long().sum()
        nll_loss = nll_loss.sum() / num_active_elements
        smoothed_loss = smoothed_loss.sum() / (num_active_elements * log_probs.shape[-1])
        return (1 - self.epsilon) * nll_loss + self.epsilon * smoothed_loss

**Comment by BenjaminBossan** [comment]
I have no experience with the transformers trainer, but it appears it's not too difficult to provide a custom loss: https://huggingface.co/docs/transformers/main_classes/trainer

Anyway, that's out of scope for PEFT.

**Comment by jarork** [comment]
Many thanks mate
